coremltools/converters/mil/frontend/torch/ops.py

```diff
@@ -3759,39 +3765,190 @@ def _internal_op_tensor_inplace_fill(context, node):
+    exist_slice = False
+    for index in indices:
+        if index is None:
+            exist_slice = True
+            break
+    if exist_slice:
+        # We have 2 ways to translate such torch.index_put, both have pros and cons
+        # 1. mb.scatter_nd
+        #    * pro: can handle accumulate or update
+        #    * con: can only have whole slice at the endmost dimensions
+        # 2. mb.torch_tensor_assign
+        #    * pro: can have whole slice at arbitrary dimension
+        #    * con: can only handle update
+        # Here we use mb.torch_tensor_assign
+        # TODO: explore how can we cover as many torch.index_put cases as possible
+        if accumulate:
+            raise NotImplementedError("If there exists whole slices, only update mode handled yet")
+
+        begin = [0] * x.rank
+        end = list(x.shape)
+        stride = [1] * x.rank
+        begin_mask = [True] * x.rank
+        end_mask = [True] * x.rank
+        # note: in torch slice, an indexed dim becomes size 1, rather than squeezed
+        is_dim_size1 = [False] * x.rank
+        for dim, index in enumerate(indices):
+            if index is not None:
+                if len(index.shape) > 0:
+                    index = mb.squeeze(x=index)
+                begin[dim] = index
+                end[dim] = mb.add(x=index, y=1)
+                begin_mask[dim] = False
+                end_mask[dim] = False
+                is_dim_size1[dim] = True
+        begin = mb.concat(values=begin, axis=0)
+        end = mb.concat(values=end, axis=0)
+
+        expected_values_shape = []
+        for dim in range(x.rank):
+            expected_values_shape.append(1 if is_dim_size1[dim] else x.shape[dim])
+        expected_values_shape = tuple(expected_values_shape)
+
+        if values.shape != expected_values_shape:
+            values = _broadcast(values.name + "_broadcasted", values, expected_values_shape)
+
+        updated_x = _translate_torch_tensor_assign(
+            x=x,
+            updates=values,
+            begin=begin,
+            end=end,
+            stride=stride,
+            begin_mask=begin_mask,
+            end_mask=end_mask,
+            squeeze_mask=[False] * x.rank,
+            name=node.name,
+        )
+        context.add(updated_x)
+        return
 
+    indices_type = indices[0].sym_type.get_primitive()
     if types.is_bool(indices_type):
+        # indices
         assert len(indices) == 1, "Unsupported index_put_ usage."
         indices = indices[0]
         assert (
             indices.shape == x.shape
         ), "indices shape must equal to input shape for index put operation."
         indices = mb.cast(x=indices, dtype="int32")
         indices = mb.non_zero(x=indices)
-
-    if types.is_int(indices_type):
+        # values
+        if len(values.shape) == 0:
+            values = mb.expand_dims(x=values, axes=[0])
+        if np.prod(values.shape) <= 1:
```

Comment:
Great! I am curious about which newly added unittest corresponding to this fix

Reply:
Diff is a bit ruined 😓 This is not my fix, but the original implementation
```
    if len(values.shape) == 0:
         values = mb.expand_dims(x=values, axes=[0])

     if values.rank == 1 and values.shape[0] == 1:
         reps = value_at(mb.shape(x=indices), 0)
         reps = mb.expand_dims(x=reps, axes=[0])
         values = mb.tile(x=values, reps=reps)
```

Reply:
The original logic is:
```
    if types.is_bool(indices_type):
        assert len(indices) == 1, "Unsupported index_put_ usage."
        indices = indices[0]
        assert (
            indices.shape == x.shape
        ), "indices shape must equal to input shape for index put operation."
        indices = mb.cast(x=indices, dtype="int32")
        indices = mb.non_zero(x=indices)

    if types.is_int(indices_type):
        if len(indices) > 1:
            indices = mb.stack(values=indices, axis=indices[0].rank)
        else:
            indices = mb.expand_dims(x=indices[0], axes=[-1])

    if len(values.shape) == 0:
        values = mb.expand_dims(x=values, axes=[0])

    if values.rank == 1 and values.shape[0] == 1:
        reps = value_at(mb.shape(x=indices), 0)
        reps = mb.expand_dims(x=reps, axes=[0])
        values = mb.tile(x=values, reps=reps)
```
and now it is:
```
    if types.is_bool(indices_type):
        # indices
        assert len(indices) == 1, "Unsupported index_put_ usage."
        indices = indices[0]
        assert (
            indices.shape == x.shape
        ), "indices shape must equal to input shape for index put operation."
        indices = mb.cast(x=indices, dtype="int32")
        indices = mb.non_zero(x=indices)
        # values
        if values.shape == ():
            values = mb.expand_dims(x=values, axes=[0])
        if values.rank == 1 and values.shape[0] == 1:
            reps = value_at(mb.shape(x=indices), 0)
            reps = mb.expand_dims(x=reps, axes=[0])
            values = mb.tile(x=values, reps=reps)
    elif types.is_int(indices_type):
        # indices
        if len(indices) > 1:
            indices = mb.stack(values=indices, axis=indices[0].rank)
        else:
            indices = mb.expand_dims(x=indices[0], axes=[-1])
        # values
        expected_values_shape = indices.shape[:-1] + x.shape[indices.shape[-1] :]
        if values.shape != expected_values_shape:
            values = _broadcast(values.name + "_broadcasted", values, expected_values_shape)
```
so basically the part of
```
    if len(values.shape) == 0:
        values = mb.expand_dims(x=values, axes=[0])

    if values.rank == 1 and values.shape[0] == 1:
        reps = value_at(mb.shape(x=indices), 0)
        reps = mb.expand_dims(x=reps, axes=[0])
        values = mb.tile(x=values, reps=reps)
```
only work under `types.is_bool(indices_type)`?
Do you replace the logic for the `types.is_int(indices_type)` by this ` values = _broadcast(values.name + "_broadcasted", values, expected_values_shape)`?

Reply:
Right, yeah I should clarify a bit:
1. For bool, it still has the original logic
2. For int, now it uses the new `expected_values_shape` and `_broadcast` logic

So:
1. for bool, same logic, same test cases
2. for int, new logic, added new test cases